### DISCUSSION TITLE
I made a tree-sitter grammar

### DISCUSSION THREAD

**Discussion by applejag** [deepnoodle-ai/risor#125] [Show and tell]
Wanted syntax highligh in neovim, so I hacked together a tree-sitter grammar: https://github.com/applejag/tree-sitter-risor

:)

**Comment by myzie** [comment]
Neat! I'm not well-versed in this area - tree-sitter is new to me. Tree sitter in general in usable in multiple editors, is that right?

What I had done so far was the _very_ rudimentary grammar here:
https://github.com/risor-io/risor/blob/main/vscode/syntaxes/risor.grammar.json

Syntax highlighting for non-VSCode situations is definitely needed, so thank you for working on this.

**Reply by applejag** [reply]
That's right! Editor support includes:

- Neovim (via first-party [nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter) plugin)
- Helix (built-in support <https://helix-editor.com/>)
- Emacs (built-in support since v29+)
- Lapce (built-in support <https://lapce.dev/>)
- Atom (discontinued editor, but it was an Atom maintainer who invented tree-sitter)
- Zed (sequel to Atom, built-in support <https://zed.dev/>)
- Visual Studio Code (via third-party **preview** and **deprecated** [tree-sitter](https://marketplace.visualstudio.com/items?itemName=georgewfraser.vscode-tree-sitter) extension, or first-party **preview** extension ["anycode"](https://marketplace.visualstudio.com/items?itemName=ms-vscode.anycode))

Some editors that are sadly abscent from that list:

- Visual Studio
- The whole JetBrains editor collection (talk of adding support here: <https://youtrack.jetbrains.com/issue/KTIJ-18678/Implement-instantaneous-non-semantic-syntax-highlighting>)
- Sublime text (Sublime HQ don't seem interested to add this at all)
- Notepad++
- Eclipse

(Though for editors like IntelliJ, VS, and Eclipse, an extension could make use of tree-sitter in the background. They're just lacking the support on editor-level, either built in or via some extension for general tree-sitter use.)

Tree-sitter isn't super plug-n-play. To get it into your editor, the maintainer must add it separately to each editor's repos. But as tree-sitter isn't bound to line-based regexes, it is a very stable syntax highlighting tool.

Could also be fun trying to implement an LSP using tree-sitter grammar, as the tree-sitter bindings would allow the LSP to know when only sections of the code changes, making it more efficient on larger files.

**Comment by alexaandru** [comment]
Thank you so much for this!!!

I now have pretty highlight for Risor in my own (not yet released) text editor:
![Pasted image 1](https://github.com/risor-io/risor/assets/85237/83032c1b-a18e-4b15-904f-8441a29df4cc)

`TreeSitter` and `LSP` are the main reasons I am writing my own editor in the 1st place: you can do soooo much with so little.

Have an awesome day everyone, I sure have one, now that I have such nice highlighting for Risor scripts :-)

**Comment by rubiojr** [comment]
Thank you! I couldn't find a Zed extension, so I've published one at https://github.com/rubiojr/zed-risor using your grammar.
